tensorflow/python/estimator/canned/tensor_forest.py

```diff
@@ -0,0 +1,174 @@
+])
+
+VARIANCE_PREDICTION_KEY = 'prediction_variance'
+
+
+class RandomForestGraphs(object):
+  """Builds TF graphs for random forest training and inference."""
+
+  def __init__(self,
+               params,
+               configs,
+               tree_configs=None):
+    self._params = params
+    self._configs = configs
+    self._variables = tensor_forest_ops.ForestVariables(
+        self._params,
+        tree_configs=tree_configs)
+
+  def inference_graph(self, dense_features):
+    """ Builds inference graphs"""
+
+    logits = [
+        gen_tensor_forest_ops.tensor_forest_tree_predict(
+            tree_variable,
+            dense_features,
+            self._params.logits_dimension)
+        for tree_variable in self._variables
+    ]
+
+    # shape of all_predict should be [batch_size, n_trees, logits_dimension]
+    all_predict = array_ops.stack(logits, axis=1)
+    average_values = math_ops.div(
+        math_ops.reduce_sum(all_predict, 1),
+        self._params.n_trees,
+        name='probabilities')
+
+    expected_squares = math_ops.div(
+        math_ops.reduce_sum(all_predict * all_predict, 1),
+        self._params.n_trees)
+    regression_variance = math_ops.maximum(
+        0., expected_squares - average_values * average_values)
+
+    return average_values, regression_variance
+
+  def average_size(self):
+    sizes = [gen_tensor_forest_ops.tensor_forest_tree_size(tree_variable)
+             for tree_variable in self._variables]
+    return math_ops.reduce_mean(math_ops.to_float(array_ops.stack(sizes)))
+
+
+def _tf_model_fn(features,
+                 labels,
+                 mode,
+                 head,
+                 sorted_feature_columns,
+                 forest_hparams,
+                 config,
+                 name='tensor_forest'):
+  """tensor forest model function"""
+  with ops.name_scope(name):
+    graph_builder = RandomForestGraphs(
+        forest_hparams, config)
+    # pylint: disable=protected-access
+    transformed_features = feature_column_lib._transform_features(
+        features, sorted_feature_columns)
+
+    dense_features = array_ops.concat(transformed_features.values(), axis=1)
+
+    logits, regression_variance = graph_builder.inference_graph(
+        dense_features)
+
+    summary.scalar('average_tree_size', graph_builder.average_size())
+
+  training_graph = None
+
+  def _train_op_fn(unused_loss):
+    del unused_loss
+    return training_graph
+
+  estimator_spec = head.create_estimator_spec(
+      features=features,
+      mode=mode,
+      labels=labels,
+      train_op_fn=_train_op_fn,
+      logits=logits)
+
+  extra_predictions = {
+      VARIANCE_PREDICTION_KEY: regression_variance
+  }
+
+  estimator_spec = estimator_spec._replace(
+      predictions=estimator_spec.predictions.update(extra_predictions))
+
+  return estimator_spec
+
+
+# @estimator_export('estimator.TensorForestClassifier')
+class TensorForestClassifier(estimator.Estimator):
```

Comment:
what about regressor? or in next pr?

Reply:
yeah, that's the plan, i have a tracking issue for this https://github.com/tensorflow/tensorflow/issues/21830